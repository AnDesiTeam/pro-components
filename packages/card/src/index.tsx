import type { CheckCardGroupProps, CheckCardProps } from './components/CheckCard';
import CheckCard from './components/CheckCard';
import type { StatisticProps } from './components/Statistic';
import Statistic from './components/Statistic';
import type { StatisticCardProps, StatisticsCardProps } from './components/StatisticCard';
import StatisticCard from './components/StatisticCard';
import type { ProCardProps } from './ProCard';
import ProCard from './ProCard';
import type { ProCardTabsProps } from './typing';
import type { ResultCardProps } from './components/ResultCard'
import  ResultCard from './components/ResultCard'

import 'antd/lib/card/style';

export type {
  ProCardTabsProps,
  ProCardProps,
  StatisticCardProps,
  StatisticsCardProps,
  CheckCardGroupProps,
  CheckCardProps,
  StatisticProps,
  ResultCardProps,
};
export { StatisticCard, Statistic, CheckCard, ProCard, ResultCard};

export default ProCard;
