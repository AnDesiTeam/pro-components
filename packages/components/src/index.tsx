export * from '@ant-design/pro-card';
export * from '@ant-design/pro-descriptions';
export * from '@ant-design/pro-field';
export * from '@ant-design/pro-form';
export * from '@ant-design/pro-layout';
export * from '@ant-design/pro-list';
export * from '@ant-design/pro-provider';
export * from '@ant-design/pro-skeleton';
export * from '@ant-design/pro-table';
export * from '@ant-design/pro-utils';

export { default as ProCard } from '@ant-design/pro-card'
export { default as ProDescriptions } from '@ant-design/pro-descriptions';
export { default as ProField } from '@ant-design/pro-field';
export { default as ProForm } from '@ant-design/pro-form';
export { default as ProLayout } from '@ant-design/pro-layout';
export { default as ProList } from '@ant-design/pro-list';
export { default as ProProvider } from '@ant-design/pro-provider';
export { default as ProSkeleton } from '@ant-design/pro-skeleton';
export { default as ProTable } from '@ant-design/pro-table';

// @ts-ignore
export * from './version';
