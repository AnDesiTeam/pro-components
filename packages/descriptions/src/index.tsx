import React, { useEffect } from 'react';
import { Descriptions, Space } from 'antd';
import toArray from 'rc-util/lib/Children/toArray';
import Field, {
  ProFieldValueType,
  ProFieldValueEnumMap,
  ProFieldValueEnumObj,
  ProFieldFCMode,
  ProRenderFieldProps,
} from '@ant-design/pro-field';
import { LabelIconTip } from '@ant-design/pro-utils';
import get from 'rc-util/lib/utils/get';
import { stringify } from 'use-json-comparison';
import { ProColumnType } from '@ant-design/pro-table';
import { DescriptionsItemProps } from 'antd/lib/descriptions/Item';
import { DescriptionsProps } from 'antd/lib/descriptions';
import useFetchData, { RequestData } from './useFetchData';

export type ActionType = {
  reload: () => void;
};

export type ProDescriptionsProps<T = {}> = DescriptionsProps & {
  /**
   * params 参数
   * params 改变的时候会触发 reload
   * todo
   */
  params?: { [key: string]: any };
  /**
   * 获取数据的方法
   */
  request?: (params: { [key: string]: any }) => Promise<RequestData<T>>;

  columns?: Omit<ProColumnType<T>, 'renderFormItem'>[];

  /**
   * 一些简单的操作
   */
  actionRef?: React.MutableRefObject<ActionType | undefined>;
};

export type ProDescriptionsItemProps = Omit<DescriptionsItemProps, 'children'> &
  ProRenderFieldProps & {
    valueType?: ProFieldValueType;
    // 隐藏这个字段，是个语法糖，方便一下权限的控制
    hide?: boolean;
    plain?: boolean;
    mode?: ProFieldFCMode;
    /**
     * 从全局数据中取数据的key
     */
    dataIndex?: string;
    // dataIndex?: React.ReactText | React.ReactText[];

    /**
     * 映射值的类型
     */
    valueEnum?: ProFieldValueEnumMap | ProFieldValueEnumObj;
    /**
     * 远程获取枚举值
     */
    request?: () => Promise<any>;

    children?: React.ReactNode;

    tip?: string;
  };

const ProDescriptionsItem: React.FC<ProDescriptionsItemProps> = props => {
  return <Descriptions.Item {...props}>{props.children}</Descriptions.Item>;
};

const ProDescriptions = <T, U>(props: ProDescriptionsProps<T>) => {
  const { request, columns, params = {}, actionRef, onRequestError, ...rest } = props;

  const action = useFetchData<RequestData<T>>(
    async () => {
      const data = request ? await request(params) : { data: {} };
      return data;
    },
    {
      onRequestError,
      effects: [stringify(params)],
    },
  );

  /**
   * 支持 reload 的功能
   */
  useEffect(() => {
    if (actionRef) {
      actionRef.current = { reload: action.reload };
    }
  }, [action]);

  if (request && columns && !props.children) {
    const { dataSource = {} } = action;
    const options: JSX.Element[] = [];
    const dom = columns.map((itemColumn, index) => {
      const { dataIndex, render } = itemColumn;
      const data = Array.isArray(dataIndex)
        ? get(dataSource, dataIndex as string[])
        : dataSource[dataIndex as string];

      //  如果 valueType 是 option 的话，应该删除掉
      if (itemColumn.valueType === 'option') {
        options.push(
          <Field
            mode="read"
            render={text => (render ? render(text, data, index) : text)}
            valueType="option"
            text={data}
          />,
        );
        return null;
      }
      // title 一定要存在，不然就不展示，因为我们不支持嵌套
      if (!itemColumn.title) {
        return null;
      }
      //  dataIndex 无所谓是否存在
      // 有些时候不需要 dataIndex 可以直接 render
      const valueType =
        typeof itemColumn.valueType === 'function'
          ? itemColumn.valueType(data)
          : itemColumn.valueType;

      return (
        <Descriptions.Item
          key={Array.isArray(dataIndex) ? dataIndex.join('.') : dataIndex}
          label={<LabelIconTip label={itemColumn.title} tip={itemColumn.tip} />}
        >
          <Field
            valueEnum={itemColumn.valueEnum}
            mode="read"
            render={text => (render ? render(text, data, index) : text)}
            valueType={valueType}
            // request={itemColumn.request}
            text={data || itemColumn.title}
          />
        </Descriptions.Item>
      );
    });

    return (
      <Descriptions extra={<Space>{options}</Space>} {...rest}>
        {dom}
      </Descriptions>
    );
  }

  const options: JSX.Element[] = [];
  // 因为 Descriptions 只是个语法糖，children 是不会执行的，所以需要这里处理一下
  const mergeChildren = toArray(props.children).map((item, index) => {
    const {
      valueType,
      children,
      valueEnum,
      render,
      mode,
      request: requestItem,
      renderFormItem,
      plain,
      dataIndex,
      ...restItem
    } = item.props as ProDescriptionsItemProps;
    if (!valueType && !valueEnum && !request) {
      return item;
    }

    if (request && !columns && dataIndex) {
      const { dataSource = {} } = action;
      return (
        <Descriptions.Item
          {...restItem}
          key={restItem.label?.toString() || index}
          label={<LabelIconTip label={restItem.label} tip={restItem.tip} />}
        >
          <Field
            valueEnum={valueEnum}
            mode={mode || 'read'}
            render={render}
            renderFormItem={renderFormItem}
            valueType={valueType}
            plain={plain}
            request={requestItem}
            text={children || dataSource[dataIndex] || restItem.label}
          />
        </Descriptions.Item>
      );
    }

    if (valueType === 'option') {
      options.push(
        <Field
          valueEnum={valueEnum}
          mode={mode || 'read'}
          render={render}
          renderFormItem={renderFormItem}
          valueType={valueType}
          plain={plain}
          text={children as string}
        />,
      );
      return null;
    }

    const field = (
      <Descriptions.Item
        {...restItem}
        key={restItem.label?.toString() || index}
        label={<LabelIconTip label={restItem.label} tip={restItem.tip} />}
      >
        <Field
          valueEnum={valueEnum}
          mode={mode || 'read'}
          render={render}
          renderFormItem={renderFormItem}
          valueType={valueType}
          plain={plain}
          request={requestItem}
          text={children as string}
        />
      </Descriptions.Item>
    );
    return field;
  });
  return (
    <Descriptions extra={<Space>{options}</Space>} {...rest}>
      {mergeChildren}
    </Descriptions>
  );
};

ProDescriptions.Item = ProDescriptionsItem;

export default ProDescriptions;
