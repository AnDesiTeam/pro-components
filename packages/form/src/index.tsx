import ProFormDatePicker from './components/datePicker';
import ProFormText from './components/text';

import ProForm from './layouts/ProForm';
import QueryFilter from './layouts/QueryFilter';

export { ProFormDatePicker, ProFormText, QueryFilter };

export default ProForm;
