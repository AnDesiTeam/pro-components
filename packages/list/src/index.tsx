import React, { useMemo, useContext } from 'react';
import { List } from 'antd';
import { ListProps } from 'antd/lib/list';
import { ProFieldValueType } from '@ant-design/pro-field';
import classNames from 'classnames';
import ProTable, { ProTableProps, ProColumns } from '@ant-design/pro-table';
import { ColumnType } from 'antd/es/table';
import { ConfigContext as AntdConfigContext } from 'antd/lib/config-provider';
import get from 'rc-util/lib/utils/get';
import ProListItem from './Item';

import './index.less';

type AntdListProps<RecordType> = Omit<ListProps<RecordType>, 'rowKey'>;

export interface ProListMeta<T = any> {
  dataIndex?: string | string[];
  valueType?: ProFieldValueType;
  render?: ProColumns['render'];
}

export interface ProListMetas {
  type?: ProListMeta;
  title?: ProListMeta;
  subTitle?: ProListMeta;
  description?: ProListMeta;
  avatar?: ProListMeta;
  actions?: ProListMeta;
}

export interface ProListProps<RecordType, U extends { [key: string]: any }>
  extends Pick<
      ProTableProps<RecordType, U>,
      'dataSource' | 'loading' | 'toolBarRender' | 'rowKey' | 'headerTitle' | 'options' | 'search'
    >,
    AntdListProps<RecordType> {
  metas?: ProListMetas;
}

export type Key = React.Key;

export type TriggerEventHandler<RecordType> = (record: RecordType) => void;

const PRO_LIST_KEYS = ['title', 'subTitle', 'avatar', 'description', 'actions'];

function ProList<RecordType = any, U = any>(props: ProListProps<RecordType, U>) {
  const {
    metas,
    split,
    pagination,
    size,
    footer,
    rowKey,
    className,
    options = false,
    search = false,
    ...rest
  } = props;

  const { getPrefixCls } = useContext(AntdConfigContext);

  const columns: ProColumns[] = useMemo(() => {
    const ret: ProColumns[] = [];
    PRO_LIST_KEYS.forEach((key) => {
      if (!metas || !metas[key]) {
        return;
      }
      const meta = metas[key];
      let valueType = meta.valueType;
      if (!valueType) {
        // 给默认的 valueType
        if (key === 'avatar') {
          valueType = 'avatar';
        }
      }
      ret.push({
        key,
        ...meta,
        valueType,
      });
    });
    return ret;
  }, [metas]);
  const tableViewRender: ProTableProps<RecordType, any>['tableViewRender'] = ({
    dataSource,
    columns,
  }) => {
    return (
      <List<RecordType>
        size={size}
        footer={footer}
        split={split}
        dataSource={dataSource}
        renderItem={(item, index) => {
          const listItemProps = {};
          columns?.forEach((column: ColumnType<RecordType>) => {
            PRO_LIST_KEYS.forEach((key) => {
              if (column.key === key) {
                const dataIndex = column.dataIndex || key;
                const rawData = Array.isArray(dataIndex)
                  ? get(item, dataIndex as string[])
                  : item[dataIndex];
                listItemProps[key] = column.render ? column.render(rawData, item, index) : rawData;
              }
            });
          });
          return <ProListItem {...listItemProps} />;
        }}
      />
    );
  };

  return (
    <ProTable<RecordType>
      {...rest}
      search={search}
      options={options}
      className={classNames(getPrefixCls('pro-list'), className)}
      columns={columns}
      rowKey={rowKey}
      tableViewRender={tableViewRender}
    />
  );
}

export default ProList;
