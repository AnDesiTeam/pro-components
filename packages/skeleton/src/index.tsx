import React from 'react';
import ListPageSkeleton, { ListPageSkeletonProps } from './component/list';

const PageSkeleton: React.FC<
  ListPageSkeletonProps & {
    type?: 'list';
    active?: boolean;
  }
> = ({ type = 'list', ...rest }) => {
  if (type === 'list') {
    return <ListPageSkeleton {...rest} />;
  }
  return null;
};

export { ListPageSkeleton };

export default PageSkeleton;
