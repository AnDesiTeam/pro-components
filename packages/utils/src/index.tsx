import LabelIconTip from './components/LabelIconTip';
import FilterDropdown from './components/FilterDropdown';
import FieldLabel from './components/FieldLabel';
import InlineErrorFormItem from './components/InlineErrorFormItem';

import isBrowser from './isBrowser';
import isImg from './isImg';
import isUrl from './isUrl';
import isNil from './isNil';
import isDropdownValueType from './isDropdownValueType';
import pickProProps from './pickProProps';
import omitUndefined from './omitUndefined';
import omitUndefinedAndEmptyArr from './omitUndefinedAndEmptyArr';
import pickProFormItemProps from './pickProFormItemProps';
import useEditableArray, {
  RowEditableConfig,
  RowEditableType,
  UseEditableType,
  UseEditableUtilType,
} from './useEditableArray';
import useEditableMap, { UseEditableMapType, UseEditableMapUtilType } from './useEditableMap';

/**
 * hooks
 */
import useDebounceFn from './hooks/useDebounceFn';
import usePrevious from './hooks/usePrevious';
import conversionSubmitValue from './conversionSubmitValue';
import transformKeySubmitValue from './transformKeySubmitValue';
import parseValueToMoment from './parseValueToMoment';
import useDeepCompareEffect from './hooks/useDeepCompareEffect';
import useDocumentTitle from './hooks/useDocumentTitle';

/**
 * type
 */
import {
  ProSchema,
  ProSchemaValueEnumMap,
  ProSchemaValueEnumObj,
  ProSchemaComponentTypes,
  ProCoreActionType,
  SearchTransformKeyFn,
  ProTableEditableFnType,
} from './typing';

export type {
  ProSchema,
  ProCoreActionType,
  ProSchemaComponentTypes,
  ProSchemaValueEnumMap,
  ProSchemaValueEnumObj,
  SearchTransformKeyFn,
  ProTableEditableFnType,
  RowEditableConfig,
  RowEditableType,
  UseEditableType,
  UseEditableUtilType,
  UseEditableMapType,
  UseEditableMapUtilType,
};

export {
  LabelIconTip,
  FilterDropdown,
  FieldLabel,
  InlineErrorFormItem,
  // function
  transformKeySubmitValue,
  conversionSubmitValue,
  parseValueToMoment,
  useDocumentTitle,
  isImg,
  isNil,
  isDropdownValueType,
  omitUndefined,
  omitUndefinedAndEmptyArr,
  pickProFormItemProps,
  isUrl,
  isBrowser,
  usePrevious,
  useDebounceFn,
  pickProProps,
  useDeepCompareEffect,
  // hooks
  useEditableArray,
  useEditableMap,
};
