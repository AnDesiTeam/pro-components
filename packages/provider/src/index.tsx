//@ts-ignore
import type { Theme } from '@ant-design/cssinjs';
import { useCacheToken } from '@ant-design/cssinjs';
import { ConfigProvider as AntdConfigProvider } from 'antd';
import zh_CN from 'antd/es/locale/zh_CN';
import React, { useContext, useEffect, useMemo, useRef } from 'react';
import { SWRConfig, useSWRConfig } from 'swr';
import arEG from './locale/ar_EG';
import caES from './locale/ca_ES';
import deDE from './locale/de_DE';
import enGB from './locale/en_GB';
import enUS from './locale/en_US';
import esES from './locale/es_ES';
import faIR from './locale/fa_IR';
import frFR from './locale/fr_FR';
import hrHR from './locale/hr_HR';
import idID from './locale/id_ID';
import itIT from './locale/it_IT';
import jaJP from './locale/ja_JP';
import koKR from './locale/ko_KR';
import mnMN from './locale/mn_MN';
import msMY from './locale/ms_MY';
import plPL from './locale/pl_PL';
import ptBR from './locale/pt_BR';
import ruRU from './locale/ru_RU';
import srRS from './locale/sr_RS';
import trTR from './locale/tr_TR';
import viVN from './locale/vi_VN';
import zhCN from './locale/zh_CN';
import zhTW from './locale/zh_TW';
import type { DeepPartial, ProTokenType } from './typing/layoutToken';
import { getLayoutDesignToken } from './typing/layoutToken';
import type { ProAliasToken } from './useStyle';
import { proTheme } from './useStyle';
import { emptyTheme, defaultToken } from './useStyle/token';
import { merge } from './utils/merge';

export * from './useStyle';

export { DeepPartial, ProTokenType };

export type ProSchemaValueEnumType = {
  /** @name 演示的文案 */
  text: React.ReactNode;

  /** @name 预定的颜色 */
  status?: string;
  /** @name 自定义的颜色 */
  color?: string;
  /** @name 是否禁用 */
  disabled?: boolean;
};

/**
 * 支持 Map 和 Object
 *
 * @name ValueEnum 的类型
 */
type ProSchemaValueEnumMap = Map<React.ReactText, ProSchemaValueEnumType | React.ReactNode>;

type ProSchemaValueEnumObj = Record<string, ProSchemaValueEnumType | React.ReactNode>;

export type BaseProFieldFC = {
  /** 值的类型 */
  text: React.ReactNode;
  /** 放置到组件上 props */
  fieldProps?: any;
  /**
   * 组件的渲染模式类型
   * @option read 渲染只读模式
   * @option edit 渲染编辑模式
   * */
  mode: ProFieldFCMode;
  /**
   * 简约模式
   */
  plain?: boolean;
  /** 轻量模式 */
  light?: boolean;
  /** Label */
  label?: React.ReactNode;
  /** 映射值的类型 */
  valueEnum?: ProSchemaValueEnumObj | ProSchemaValueEnumMap;
  /** 唯一的key，用于网络请求 */
  proFieldKey?: React.Key;
};

export type ProFieldFCMode = 'read' | 'edit' | 'update';

/** Render 第二个参数，里面包含了一些常用的参数 */
export type ProFieldFCRenderProps = {
  mode?: ProFieldFCMode;
  readonly?: boolean;
  placeholder?: string | string[];
  value?: any;
  onChange?: (...rest: any[]) => void;
} & BaseProFieldFC;

export type ProRenderFieldPropsType = {
  /**
   * 自定义只读模式的渲染器
   * @params props 关于dom的配置
   * @params dom 默认的 dom
   * @return 返回一个用于读的 dom
   */
  render?:
    | ((
        text: any,
        props: Omit<ProFieldFCRenderProps, 'value' | 'onChange'>,
        dom: JSX.Element,
      ) => JSX.Element)
    | undefined;
  /**
   * 一个自定义的编辑渲染器。
   * @params text 默认的值类型
   * @params props 关于dom的配置
   * @params dom 默认的 dom
   * @return 返回一个用于编辑的dom
   */
  renderFormItem?:
    | ((text: any, props: ProFieldFCRenderProps, dom: JSX.Element) => JSX.Element)
    | undefined;
};

export type IntlType = {
  locale: string;
  getMessage: (id: string, defaultMessage: string) => string;
};

/**
 * 安全的从一个对象中读取相应的值
 * @param source
 * @param path
 * @param defaultValue
 * @returns
 */
function get(
  source: Record<string, unknown>,
  path: string,
  defaultValue?: string,
): string | undefined {
  // a[3].b -> a.3.b
  const paths = path.replace(/\[(\d+)\]/g, '.$1').split('.');
  let result = source;
  let message = defaultValue;
  // eslint-disable-next-line no-restricted-syntax
  for (const p of paths) {
    message = Object(result)[p];
    result = Object(result)[p];
    if (message === undefined) {
      return defaultValue;
    }
  }
  return message;
}

/**
 * 创建一个国际化的操作函数
 *
 * @param locale
 * @param localeMap
 */
export const createIntl = (locale: string, localeMap: Record<string, any>): IntlType => ({
  getMessage: (id: string, defaultMessage: string) =>
    get(localeMap, id, defaultMessage) || defaultMessage,
  locale,
});

const mnMNIntl = createIntl('mn_MN', mnMN);
const arEGIntl = createIntl('ar_EG', arEG);
const zhCNIntl = createIntl('zh_CN', zhCN);
const enUSIntl = createIntl('en_US', enUS);
const enGBIntl = createIntl('en_GB', enGB);
const viVNIntl = createIntl('vi_VN', viVN);
const itITIntl = createIntl('it_IT', itIT);
const jaJPIntl = createIntl('ja_JP', jaJP);
const esESIntl = createIntl('es_ES', esES);
const caESIntl = createIntl('ca_ES', caES);
const ruRUIntl = createIntl('ru_RU', ruRU);
const srRSIntl = createIntl('sr_RS', srRS);
const msMYIntl = createIntl('ms_MY', msMY);
const zhTWIntl = createIntl('zh_TW', zhTW);
const frFRIntl = createIntl('fr_FR', frFR);
const ptBRIntl = createIntl('pt_BR', ptBR);
const koKRIntl = createIntl('ko_KR', koKR);
const idIDIntl = createIntl('id_ID', idID);
const deDEIntl = createIntl('de_DE', deDE);
const faIRIntl = createIntl('fa_IR', faIR);
const trTRIntl = createIntl('tr_TR', trTR);
const plPLIntl = createIntl('pl_PL', plPL);
const hrHRIntl = createIntl('hr_', hrHR);

const intlMap = {
  'mn-MN': mnMNIntl,
  'ar-EG': arEGIntl,
  'zh-CN': zhCNIntl,
  'en-US': enUSIntl,
  'en-GB': enGBIntl,
  'vi-VN': viVNIntl,
  'it-IT': itITIntl,
  'ja-JP': jaJPIntl,
  'es-ES': esESIntl,
  'ca-ES': caESIntl,
  'ru-RU': ruRUIntl,
  'sr-RS': srRSIntl,
  'ms-MY': msMYIntl,
  'zh-TW': zhTWIntl,
  'fr-FR': frFRIntl,
  'pt-BR': ptBRIntl,
  'ko-KR': koKRIntl,
  'id-ID': idIDIntl,
  'de-DE': deDEIntl,
  'fa-IR': faIRIntl,
  'tr-TR': trTRIntl,
  'pl-PL': plPLIntl,
  'hr-HR': hrHRIntl,
};

const intlMapKeys = Object.keys(intlMap);

export type ParamsType = Record<string, any>;

export {
  mnMNIntl,
  arEGIntl,
  enUSIntl,
  enGBIntl,
  zhCNIntl,
  viVNIntl,
  itITIntl,
  jaJPIntl,
  esESIntl,
  caESIntl,
  ruRUIntl,
  srRSIntl,
  msMYIntl,
  zhTWIntl,
  frFRIntl,
  ptBRIntl,
  koKRIntl,
  idIDIntl,
  deDEIntl,
  faIRIntl,
  trTRIntl,
  plPLIntl,
  hrHRIntl,
  intlMap,
  intlMapKeys,
};

/**
 * 自带的token 配置
 */
export type ConfigContextPropsType = {
  intl?: IntlType;
  valueTypeMap?: Record<string, ProRenderFieldPropsType>;
  token?: ProAliasToken;
  hashId?: string;
  hashed?: boolean;
  dark?: boolean;
  theme?: Theme<any, any>;
  containerDomRef?: React.RefObject<HTMLDivElement>;
};

/* Creating a context object with the default values. */
const ProConfigContext = React.createContext<ConfigContextPropsType>({
  intl: {
    ...zhCNIntl,
    locale: 'default',
  },
  valueTypeMap: {},
  theme: emptyTheme,
  hashed: true,
  dark: false,
  token: defaultToken as ProAliasToken,
});

export const { Consumer: ConfigConsumer } = ProConfigContext;

/**
 * 根据 antd 的 key 来找到的 locale 插件的 key
 *
 * @param localeKey
 */
const findIntlKeyByAntdLocaleKey = <T extends string | undefined>(localeKey: T) => {
  if (!localeKey) {
    return 'zh-CN' as T;
  }
  const localeName = localeKey.toLocaleLowerCase();
  return intlMapKeys.find((intlKey) => {
    const LowerCaseKey = intlKey.toLocaleLowerCase();
    return LowerCaseKey.includes(localeName);
  }) as T;
};

/**
 * 组件解除挂载后清空一下 cache
 * @date 2022-11-28
 * @returns null
 */
const CacheClean = () => {
  const { cache } = useSWRConfig();

  useEffect(() => {
    return () => {
      // is a map
      // @ts-ignore
      cache.clear();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);
  return null;
};

const ConfigProVidContainer: React.FC<{
  children: React.ReactNode;
  autoClearCache?: boolean;
  valueTypeMap?: Record<string, ProRenderFieldPropsType>;
  token?: DeepPartial<ProAliasToken>;
  hashed?: boolean;
  dark?: boolean;
}> = (props) => {
  const { children, dark, valueTypeMap, autoClearCache = false, token: propsToken } = props;
  const { locale, getPrefixCls, ...restConfig } = useContext(AntdConfigProvider.ConfigContext);
  const tokenContext = proTheme.useToken?.();
  const containerDomRef = useRef<HTMLDivElement>(null);
  const proProvide = useContext(ProConfigContext);

  /**
   * pro 的 类
   * @type {string}
   * @example .ant-pro
   */
  const proComponentsCls = `.${getPrefixCls()}-pro`;

  const antCls = '.' + getPrefixCls();

  const salt = `${proComponentsCls}`;

  /**
   * 合并一下token，不然导致嵌套 token 失效
   */
  const proLayoutTokenMerge = useMemo(() => {
    return getLayoutDesignToken(propsToken || {}, tokenContext.token || defaultToken);
  }, [propsToken, tokenContext.token]);

  const proProvideValue = useMemo(() => {
    const localeName = locale?.locale;
    const key = findIntlKeyByAntdLocaleKey(localeName);
    // antd 的 key 存在的时候以 antd 的为主
    const intl =
      localeName && proProvide.intl?.locale === 'default'
        ? intlMap[key!]
        : proProvide.intl || intlMap[key!];
    return {
      ...proProvide,
      dark: dark ?? proProvide.dark,
      token: merge(proProvide.token, tokenContext.token, {
        proComponentsCls,
        antCls,
        themeId: tokenContext.theme.id,
        layout: proLayoutTokenMerge,
      }),
      intl: intl || zhCNIntl,
    };
  }, [
    locale?.locale,
    proProvide,
    dark,
    tokenContext.token,
    tokenContext.theme.id,
    proComponentsCls,
    antCls,
    proLayoutTokenMerge,
  ]);

  const finalToken = {
    ...(proProvideValue.token || {}),
    proComponentsCls,
  };

  const [token, nativeHashId] = useCacheToken<ProAliasToken>(
    tokenContext.theme,
    [tokenContext.token, finalToken ?? {}],
    {
      salt,
    },
  );

  const hashId = useMemo(() => {
    if (props.hashed === false) {
      return '';
    }
    if (proProvide.hashed === false) return '';
    if (process.env.NODE_ENV?.toLowerCase() !== 'test') return nativeHashId;
    return '';
  }, [nativeHashId, proProvide.hashed, props.hashed]);

  const configProviderDom = useMemo(() => {
    const themeConfig = {
      ...restConfig.theme,
      hashId: hashId,
      hashed: process.env.NODE_ENV?.toLowerCase() !== 'test' && proProvide.hashed,
    };

    const provide = (
      <AntdConfigProvider {...restConfig} theme={{ ...themeConfig }}>
        <ProConfigContext.Provider
          value={{
            ...proProvideValue!,
            valueTypeMap: valueTypeMap || proProvideValue?.valueTypeMap,
            token,
            containerDomRef,
            theme: tokenContext.theme,
            hashed: props.hashed,
            hashId,
          }}
        >
          <>
            {autoClearCache && <CacheClean />}
            {children}
          </>
        </ProConfigContext.Provider>
      </AntdConfigProvider>
    );

    return (
      <div
        ref={containerDomRef}
        className={`${getPrefixCls?.('pro') || 'ant-pro'}${hashId ? ' ' + hashId : ''}`}
      >
        {provide}
      </div>
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [autoClearCache, children, getPrefixCls, hashId, locale, proProvideValue, token]);

  if (!autoClearCache) return configProviderDom;

  return <SWRConfig value={{ provider: () => new Map() }}>{configProviderDom}</SWRConfig>;
};

export const ProConfigProvider: React.FC<{
  children: React.ReactNode;
  autoClearCache?: boolean;
  token?: DeepPartial<ProAliasToken>;
  needDeps?: boolean;
  valueTypeMap?: Record<string, ProRenderFieldPropsType>;
  dark?: boolean;
  hashed?: boolean;
}> = (props) => {
  const { needDeps, dark, token } = props;
  const proProvide = useContext(ProConfigContext);
  const { locale, theme, ...rest } = useContext(AntdConfigProvider.ConfigContext);

  // 是不是不需要渲染 provide
  const isNullProvide =
    needDeps &&
    proProvide.hashId !== undefined &&
    Object.keys(props).sort().join('-') === 'children-needDeps';

  if (isNullProvide) return <>{props.children}</>;

  const mergeAlgorithm = () => {
    const isDark = dark ?? proProvide.dark;
    if (isDark && !Array.isArray(theme?.algorithm)) {
      return [proTheme.darkAlgorithm, theme?.algorithm].filter(Boolean);
    }
    if (isDark && Array.isArray(theme?.algorithm)) {
      return [proTheme.darkAlgorithm, ...(theme?.algorithm || [])].filter(Boolean);
    }
    return theme?.algorithm;
  };
  // 自动注入 antd 的配置
  const configProvider = {
    ...rest,
    locale: locale || zh_CN,
    theme: {
      ...theme,
      algorithm: mergeAlgorithm(),
    },
  } as typeof theme;

  return (
    <AntdConfigProvider {...configProvider}>
      <ConfigProVidContainer {...props} token={token} />
    </AntdConfigProvider>
  );
};

/**
 * It returns the intl object from the context if it exists, otherwise it returns the intl object for
 * the current locale
 * @returns The return value of the function is the intl object.
 */
export function useIntl(): IntlType {
  const { locale } = useContext(AntdConfigProvider.ConfigContext);
  const { intl } = useContext(ProConfigContext);

  if (intl && intl.locale !== 'default') {
    return intl || zhCNIntl;
  }

  if (locale?.locale) {
    return intlMap[findIntlKeyByAntdLocaleKey(locale.locale)] || zhCNIntl;
  }

  return zhCNIntl;
}

const ProProvider = ProConfigContext;

ProProvider.displayName = 'ProProvider';

export { ProProvider };

export default ProConfigContext;
