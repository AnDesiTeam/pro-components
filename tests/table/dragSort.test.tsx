import { mount } from 'enzyme';
import React from 'react';
import ProTable from '@ant-design/pro-table';
import { request } from './demo';
import { waitForComponentToPaint } from '../util';
import { MenuOutlined } from '@ant-design/icons';
import { sortData } from '../../packages/table/src/utils';

describe('dragSort', () => {
  it('🔥 [dragSort] render drag sort default handle by dragSortKey', async () => {
    const html = mount(
      <ProTable
        size="small"
        columns={[
          {
            title: '排序',
            dataIndex: 'sort',
          },
          {
            title: '姓名',
            dataIndex: 'name',
            className: 'drag-visible',
          },
          {
            title: '年龄',
            dataIndex: 'age',
          },
          {
            title: '地址',
            dataIndex: 'address',
          },
        ]}
        request={request}
        rowKey="key"
        dragSortKey="sort"
      />,
    );
    await waitForComponentToPaint(html, 1200);
    expect(html.find('.dragSortDefaultHandle')).toMatchSnapshot();
  });

  it('🔥 [dragSort] render drag sort custom handle by dragSortHandlerRender', async () => {
    const dragHandleRender = (rowData: any, idx: any) => (
      <>
        <MenuOutlined className="dragSortCustomHandle" style={{ cursor: 'grab', color: 'gold' }} />
        {idx + 1} - {rowData.name}
      </>
    );
    const html = mount(
      <ProTable
        size="small"
        columns={[
          {
            title: '排序',
            dataIndex: 'sort',
          },
          {
            title: '姓名',
            dataIndex: 'name',
            className: 'drag-visible',
          },
          {
            title: '年龄',
            dataIndex: 'age',
          },
          {
            title: '地址',
            dataIndex: 'address',
          },
        ]}
        request={request}
        rowKey="key"
        dragSortKey="sort"
        dragSortHandlerRender={dragHandleRender}
      />,
    );
    await waitForComponentToPaint(html, 1200);
    expect(html.find('.dragSortCustomHandle')).toMatchSnapshot();
  });

  it('🔥 [dragSort] sort core', async () => {
    type DataSourceItemStruct = {
      id: number;
      name: string;
    };
    const dataSource: DataSourceItemStruct[] = [
      {
        id: 1,
        name: 'kiner',
      },
      {
        id: 2,
        name: 'WenHui Tang',
      },
      {
        id: 3,
        name: 'Kiner Tang',
      },
    ];
    const newDs: DataSourceItemStruct[] | null = sortData<DataSourceItemStruct>(
      { oldIndex: 1, newIndex: 0 },
      dataSource,
    );
    expect(newDs).not.toBe(null);
    if (newDs) expect(newDs[0].id).toBe(2);
  });
});
