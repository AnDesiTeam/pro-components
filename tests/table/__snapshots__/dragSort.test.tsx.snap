// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`dragSort ðŸ”¥ [dragSort] custom render function 1`] = `null`;

exports[`dragSort ðŸ”¥ [dragSort] custom render function 2`] = `null`;

exports[`dragSort ðŸ”¥ [dragSort] render drag sort custom handle by dragSortHandlerRender 1`] = `null`;

exports[`dragSort ðŸ”¥ [dragSort] render drag sort default handle by dragSortKey 1`] = `null`;
